### ISSUE TITLE
Plan for Symbiont pipeline

### ISSUE THREAD

**Issue by matentzn** [monarch-initiative/mondo-ingest#27]
### Phase 1: Map: Updating the mappings

**Goal:**: The purpose of the mapping phase is to establish a as complete as possible mapping between classes in O. Technically, this involves updating M_pos and M_neg.

**Input:**
- Merged Ontology O
- Fixed positive mappings M_pos
- Fixed negative mappings M_neg
- Mapping Candidates M_c

**Process:**
1. Generate lexical mapping M_x
1. Run boomer with O, M_pos (0.99), M_c (0.8), M_x (0.5). 
   - **Output:**
      - Reconciled mapping M_r
      - New mappings M_r_n < M_r
      - Removed mappings M_r_r < M_r (there should be very few, if any)
1. Human review of M_r_n and M_r_r. 
   - If M in M_r_r is 
       - wrongly removed, affirm it with a review statement
       - rightly removed, add to M_neg
   -  If M in M_r_n is
       - wrong: Add to M_neg
       - good: Add to M_pos

**Output:**
- Updated fixed positive mappings M_pos
- Updated fixed negative mappings M_neg

### Phase 2: Migrate

**Input**:
- Source ontology O
- List of classes in source ontology: T_s
- List of classes in target ontology/ontologies: T_t
- Fixed positive mappings: M_pos

**Process**
1. Identify unmapped elements in T_t. Export as ROBOT template with label xref, definition parent.
2. Human reviewer reviews each proposed absorption and decides:
    1. Include: Merge into O, add to M_pos
    2. Exclude: Add to exclusion table
    3. Remap: HR decides that there is already a class that should be used for the mapping instead. Updating M_pos.

### Phase 3: Sync

**Input**:
- Source ontology O_s
- Target ontology/ontologies O_t
- Fixed positive mappings: M_pos
- Rejected axioms: O_reject

**Process**
1. Remove all support from O_t in all axioms in O_s
2. Import all axioms from O_t that is not in O_s or O_reject.
3. Add support for all axioms in O_s that are in O_t.
4. Human reviewer looks at all new axioms O_s (O_t-(O_s+O_reject)). 
    - If the axiom is bad, add the axiom to (O_reject)
 5. Repeat process (start from 1) until all axioms are good (if all bad axioms are removed in the first run, just 1 second run is needed).

**Comment by matentzn** [comment]
- [ ] Figure out a way to publish the Mondo paper in portions

**Comment by joeflack4** [comment]
### Questions
1. **Input: source ontology O_s**: Is this always `mondo.owl` for purposes of my script?
2. Fixed positive mappings: M_pos: Is this always `mondo.sssom.tsv` for purposes of my script?

**Comment by matentzn** [comment]
> Input: source ontology O_s: Is this always tmp/mondo.owl for purposes of my script?

I renamed the parameters now to make more sense. Mondo is the target, mondo ingest the source. I also added explanations which is which file.

> Fixed positive mappings: M_pos: Is this always tmp/mondo.sssom.tsv for purposes of my script?

Yes.

> Output: Should I save as sync/mondo-sync-<ontology>.owl for now? Just to avoid accidentally getting overwritten by tmp/mondo.owl. Ideally I would imagine we'd want to read/write from the same file; either tmp/mondo.owl or a sync/mondo-sync.owl. For first synonym step, it looks like output is just sync/synonyms.robot.template.tsv

For now, we will only ever output robot template files. Once KGCL is ready, we will output KGCL files.

> Output: Does each of the 5 steps update (and ultimately read from) the same output (e.g. sync/mondo-sync.owl), or separate outputs?

For now, to keep things simple, we will only output additive stuff. Just robot templates, nothing else.

> Recommended way to update existing .owl: ODK? robot / sparql?

No need to do that now.

> Algorithm pseudocode: It is axiom centric, but if I replaced the word 'axioms' with one of synonyms, xrefs, definitions, and obsoletion, would a lot of this algorithm apply to each of those as well?

Exactly! Axioms includes all of the ones you mention and more.
